rules/alerting.go

```diff
@@ -316,20 +315,58 @@ func (r *AlertingRule) Restored() bool {
 	return r.restored.Load()
 }
 
-func (r *AlertingRule) SetNoDependentRules(noDependentRules bool) {
-	r.noDependentRules.Store(noDependentRules)
+func (r *AlertingRule) SetDependentRules(dependents []Rule) {
+	r.dependenciesMutex.Lock()
+	defer r.dependenciesMutex.Unlock()
+
+	r.dependentRules = make(map[string]struct{})
+	for _, dependent := range dependents {
+		r.dependentRules[dependent.Name()] = struct{}{}
+	}
 }
 
 func (r *AlertingRule) NoDependentRules() bool {
-	return r.noDependentRules.Load()
+	r.dependenciesMutex.RLock()
+	defer r.dependenciesMutex.RUnlock()
+
+	if r.dependentRules == nil {
+		return false // We don't know if there are dependent rules.
```

Comment:
What's the reasoning behind this? Technically, we shouldn't ask before we populate this so I don't see the point of making this return false before the computation happens.

Reply:
This is related to https://github.com/prometheus/prometheus/pull/15560. In the case of an indeterminate condition, we exit early and return a nil dependency map.

I saw two options here:
- Differentiate between nil and an empty dependencies/dependents collection. `nil` means indeterminate (not calculated or unable to calculate), while empty means no rules
- Modify the implementation of the bugfix above to populate a full dependant/dependency collection in the case of an indeterminate condition (this was the initial implementation of the bugfix PR above, if you remember)

I went with the first option

Reply:
Right! Now I remember.